core/drivers/imx/dcp/dcp.c

```diff
@@ -0,0 +1,776 @@
+
+	return ret;
+}
+
+TEE_Result dcp_cipher_do_update(struct dcp_cipher_data *data,
+				const uint8_t *src, uint8_t *dst, size_t size)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	struct dcp_align_buf output = { };
+	struct dcp_align_buf input = { };
+	struct dcp_descriptor *desc = NULL;
+
+	if (!data || !src || !dst)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (!data->initialized) {
+		EMSG("Error, please call dcp_aes_do_init() before");
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	if (size % DCP_AES128_BLOCK_SIZE) {
+		EMSG("Input size has to be a multiple of %zu bytes",
+		     DCP_AES128_BLOCK_SIZE);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	ret = dcp_calloc_align_buf(&output, size);
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	ret = dcp_calloc_align_buf(&input, size);
+	if (ret != TEE_SUCCESS)
+		goto out;
+
+	desc = &data->dcp_data.desc;
+
+	/* Copy input data */
+	memcpy(input.data, src, size);
+
+	/* Copy key and IV */
+	memcpy(data->payload.data, data->key, DCP_AES128_KEY_SIZE);
+	data->payload_size = DCP_AES128_KEY_SIZE;
+	if (desc->ctrl0 & DCP_CONTROL0_CIPHER_INIT) {
+		memcpy(data->payload.data + DCP_AES128_KEY_SIZE, data->iv,
+		       DCP_AES128_IV_SIZE);
+		data->payload_size += DCP_AES128_IV_SIZE;
+	}
+
+	desc->src_buffer = (uint8_t *)input.paddr;
+	desc->dest_buffer = (uint8_t *)output.paddr;
+	desc->payload = (uint8_t *)data->payload.paddr;
+	desc->buff_size = size;
+
+	cache_operation(TEE_CACHECLEAN, data->payload.data,
+			data->payload_size);
+	cache_operation(TEE_CACHECLEAN, input.data, size);
+	cache_operation(TEE_CACHEINVALIDATE, output.data, size);
+
+	ret = dcp_run(&data->dcp_data);
+	if (ret)
+		goto out;
+
+	cache_operation(TEE_CACHEINVALIDATE, output.data, size);
+
+	desc->ctrl0 &= ~DCP_CONTROL0_CIPHER_INIT;
+
+	memcpy(dst, output.data, size);
+out:
+	dcp_free(&output);
+	dcp_free(&input);
+
+	return ret;
+}
+
+void dcp_cipher_do_final(struct dcp_cipher_data *data)
+{
+	if (data)
+		data->initialized = false;
+
+	dcp_free(&data->payload);
+	dcp_unlock_channel(data->dcp_data.channel);
+}
+
+TEE_Result dcp_sha_do_init(struct dcp_hash_data *hashdata)
+{
+	struct dcp_descriptor *desc = NULL;
+	TEE_Result ret = TEE_ERROR_GENERIC;
+
+	if (!hashdata) {
+		EMSG("Bad parameters, hashdata is NULL");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	desc = &hashdata->dcp_data.desc;
+
+	/* DCP descriptor init */
+	desc->status = 0;
+	desc->payload = NULL;
+	desc->dest_buffer = NULL;
+	desc->ctrl0 = DCP_CONTROL0_ENABLE_HASH | DCP_CONTROL0_INTERRUPT_ENABLE |
+		      DCP_CONTROL0_DECR_SEMAPHORE | DCP_CONTROL0_HASH_INIT;
+	desc->ctrl1 = hash_alg[hashdata->alg].type;
+	desc->buff_size = 0;
+	desc->next = 0;
+	desc->src_buffer = NULL;
+
+	ret = dcp_lock_channel(&hashdata->dcp_data.channel);
+	if (ret != TEE_SUCCESS) {
+		EMSG("Channel is busy, can't start operation now");
+		return ret;
+	}
+
+	/* Allocate context data */
+	ret = dcp_calloc_align_buf(&hashdata->ctx, DCP_SHA_BLOCK_SIZE);
+	if (ret != TEE_SUCCESS)
+		return TEE_ERROR_OUT_OF_MEMORY;
```

Comment:
Why isn't `ret` returned?

Reply:
Fixed it to return ```ret``` instead. I also fixed other similar mistake